futures-test/src/task/panic_waker.rs

```diff
@@ -1,74 +1,56 @@
-use futures_core::task::{LocalWaker, UnsafeWake, Wake, Waker};
-use std::cell::UnsafeCell;
-use std::ptr::NonNull;
-use std::sync::Arc;
+use futures_core::task::{LocalWaker, RawWaker, RawWakerVTable};
+use core::cell::UnsafeCell;
+use core::ptr::null;
 
-/// An implementation of [`Wake`](futures_core::task::Wake) that panics when
-/// woken.
-///
-/// # Examples
-///
-/// ```should_panic
-/// #![feature(futures_api)]
-/// use futures_test::task::panic_local_waker_ref;
-///
-/// let lw = panic_local_waker_ref();
-/// lw.wake(); // Will panic
-/// ```
-#[derive(Debug)]
-pub struct PanicWake {
-    _reserved: (),
+unsafe fn noop_clone(_data: *const()) -> RawWaker {
+    raw_panic_waker()
 }
 
-impl PanicWake {
-    /// Create a new instance
-    pub fn new() -> Self {
-        Self { _reserved: () }
-    }
+unsafe fn noop(_data: *const()) {
```

Comment:
```suggestion
fn noop(_data: *const()) {
```

You should be able to remove `unsafe` here because of this coercion between `fn` and `unsafe fn`:
```rust
fn noop(_data: *const()) {}
const PTR: unsafe fn(*const ()) = noop;
```

Reply:
Interesting, so there’s a coercion from `fn` -> `unsafe fn` and from environment-less-closure to `fn`, but not the combined environment-less-closure to `unsafe fn`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=eaec9c9f080c999a55a08dd2184d8669.  Has there been any discussion of adding the latter?

Reply:
We generally accept PRs for these without FCP since the original RFC specified that we should include all transitive coercions, but there isn't a mechanism in the compiler for that today.